interface/unions/README.md

```diff
@@ -0,0 +1,75 @@
+
+# Union demo (demonstration release)
+
+This project demonstrates declaration of Unions in a smithy interface file,
+and using the code generated from them. 
+
+| ____ Demo release ____                                                                                                                                                                                                                                                                                                                                     |
+|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
+| This is a demonstration release. Code generation for unions may incur breaking changes over the next 1.5 months (through end of March 2022) while we gather feedback from internal and extenral developers. Please try it out and let us know on wasmcloud Slack (or as an issue in this repo) if you want to use unions and if you find this useful. |
+
+
+Running:
+- To build, `cargo build`
+- To run the example (`examples/main.rs`): `cargo run --example main`
+- To run the tests: `cargo test -- --nocapture`
+
+## Defining unions in smithy files
+
+To define a union, use the `union` smithy type. Each variant requires a field name,
+a data type, and a `@n(_)` trait for the field number. Variant data types may be any
+smithy data type, including simple types, structures, arrays, and maps.
+
+```smithy
+union Number {
+   @n(0)
+   intVal: U64
+   
+   @n(1)
+   floatVal: F64
+}
+```
+
+Add `protocol="2"` to the `@wasmbus` trait for any service whose method parameters contain unions (directly or indirectly). You'll get an error if a union is used by a sevice that doesn't declare `protocol="2"`.
+```smithy
+@wasmbus(
+    contractId: "wasmcloud:example:union_demo",
+    actorReceive: true,
+    providerReceive: true,
+    protocol: "2" )
+service UnionDemo {
+    version: "0.1",
+    operations: [ Get ]
+}
+```
+
+Also make sure you have the latest rpc and code generation dependencies: in `Cargo.toml`:
+```toml
+[dependencies]
+wasmbus-rpc = "0.7.3"
+
+[dev-dependencies]
+weld-codegen = "0.3.2"
+
+```
+
+**Important**: Setting protocol to "2" changes message serialization from msgpack to CBOR. 
```

Comment:
Interesting, so protocol is set at the interface level so _in theory_ we shouldn't have too much trouble coordinating dependencies between actors and providers. If they use the same (semver compatible) version of an interface then they use the same serialization protocol.

I think we need to consider incorporating this into the contract ID or similar. Actors and Providers shouldn't have to agree exactly on the same version of an interface, but they will need to use the same protocol. For example, there's nothing stopping you from starting `wasmcloud.azurecr.io/echo:0.1.0` and `wasmcloud.azurecr.io/httpserver:0.14.9`, both of which are on different versions of the same interface

Reply:
> different versions of the same interface

"same interface" means a set of messages and agreement on the wire protocol

Reply:
Also,fwiw, this is consistent with the way Amazon uses protocol definitions in a smithy file. `@wasmbus` is actually declared as a protocol trait. 

https://github.com/wasmCloud/interfaces/blob/23f79b5c5e6ba9a69b0a7547daa9952b164271a1/core/wasmcloud-model.smithy#L115

Reply:
I'm coming around to the idea that we shouldn't use protocol= inside `@wasmbus` but should instead add a new protocol trait `@cbor` or `@cbor(min_version)`

That makes it more explicit

Reply:
Does this mean we can't use unions unless we also use CBOR?

Reply:
> Does this mean we can't use unions unless we also use CBOR?

Correct